### PR TITLE
Fix a low-probability crash in our qsort implementation

### PR DESCRIPTION
**Pull request by locker**
This is a backport of commit https://github.com/postgres/postgres/commit/9d6077abf9d6efd992a59f05ef5aba981ea32096.

### CODE REVIEW

File: third_party/qsort_arg_mt.c [hunk 1]
```diff
@@ -204,20 +205,42 @@ qsort_arg_mt_internal(void *a, size_t n, intptr_t es,
 		pc -= es;
 	}
 	pn = (char *) a + n * es;
-	r = min(pa - (char *) a, pb - pa);
-	vecswap((char*)a, pb - r, r);
-	r = min(pd - pc, pn - pd - es);
-	vecswap(pb, pn - r, r);
-	if ((r = pb - pa) > es) {
+	d1 = min(pa - (char *) a, pb - pa);
```
Comment (alyapunov) [inline third_party/qsort_arg_mt.c:208]: Actually these four lines of diff are meaningless, the old code (r=;vecswap;r=;vecswap;) was technically the same.

**Review by alyapunov — APPROVED** [review]
I understand that this is back-porting, but actually this patch does not fix "risk of sorting INT_MAX bytes" since that bug has already been fixed in cf1c420cdff448ab6e50f77c7255264f27d05dd6.

File: third_party/qsort_arg.c [hunk 2]
```diff
@@ -197,20 +198,39 @@ loop:SWAPINIT(a, es);
...
+	d2 = pd - pc;
+	if (d1 <= d2)
 	{
-		/* Iterate rather than recurse to save stack space */
-		a = pn - r;
-		n = r / es;
-		goto loop;
+		/* Recurse on left partition, then iterate on right partition */
+		if (d1 > es)
+			qsort_arg(a, d1 / es, es, cmp, arg);
```
Comment (alyapunov) [inline third_party/qsort_arg.c:211]: Nit: actually qsort_arg_st must be here.

File: third_party/qsort_arg.c [hunk 3]
```diff
@@ -197,20 +198,39 @@ loop:SWAPINIT(a, es);
...
+			a = pn - d2;
+			n = d2 / es;
+			goto loop;
+		}
+	}
+	else
+	{
+		/* Recurse on right partition, then iterate on left partition */
+		if (d2 > es)
+			qsort_arg(pn - d2, d2 / es, es, cmp, arg);
```
Comment (alyapunov) [inline third_party/qsort_arg.c:225]: And here..

File: third_party/qsort_arg.c (hunk 2 above)
Comment (locker) [reply third_party/qsort_arg.c]: Right. Fixed.

File: third_party/qsort_arg.c (hunk 3 above)
Comment (locker) [reply third_party/qsort_arg.c]: Fixed.

File: third_party/qsort_arg_mt.c (hunk 1 above)
Comment (locker) [reply third_party/qsort_arg_mt.c:208]: Still, I think it's better to apply this diff so that our code doesn't diverge from the origin.

**Comment by locker** [comment]
> I understand that this is back-porting, but actually this patch does not fix "risk of sorting INT_MAX bytes" since that bug has already been fixed in [cf1c420](https://github.com/tarantool/tarantool/commit/cf1c420cdff448ab6e50f77c7255264f27d05dd6).

Removed the corresponding paragraph from the commit message.

**Comment by locker** [comment]
Cherry-picked to 2.10.

**Pull request merged**